man/ocamldep.1
.\"**************************************************************************
.\"*                                                                        *
.\"*                                 OCaml                                  *
.\"*                                                                        *
.\"*             Xavier Leroy, projet Cristal, INRIA Rocquencourt           *
.\"*                                                                        *
.\"*   Copyright 1996 Institut National de Recherche en Informatique et     *
.\"*     en Automatique.                                                    *
.\"*                                                                        *
.\"*   All rights reserved.  This file is distributed under the terms of    *
.\"*   the GNU Lesser General Public License version 2.1, with the          *
.\"*   special exception on linking described in the file LICENSE.          *
.\"*                                                                        *
.\"**************************************************************************
.\"
.TH OCAMLDEP 1

.SH NAME
ocamldep \- Dependency generator for OCaml

.SH SYNOPSIS
.B ocamldep
[
.I options
]
.I filename ...

.SH DESCRIPTION

The
.BR ocamldep (1)
command scans a set of OCaml source files
(.ml and .mli files) for references to external compilation units,
and outputs dependency lines in a format suitable for the
.BR make (1)
utility. This ensures that make will compile the source files in the
correct order, and recompile those files that need to when a source
file is modified.

The typical usage is:
.P
ocamldep
.I options
*.mli *.ml > .depend
.P
where .depend is the file that should contain the
dependencies.

Dependencies are generated both for compiling with the bytecode
compiler
.BR ocamlc (1)
and with the native-code compiler
.BR ocamlopt (1).

.SH OPTIONS

The following command-line options are recognized by
.BR ocamldep (1).
.TP
.B \-absname
Show absolute filenames in error messages.
.TP
.B \-no-absname
Do not try to show absolute filenames in error messages (default).
.TP
.B \-all
Generate dependencies on all required files, rather than assuming
implicit dependencies.
.TP
.B \-allow\-approx
Allow falling back on a lexer-based approximation when parsing fails.
.TP
.B \-as\-map
For the following files, do not include delayed dependencies for
module aliases.
This option assumes that they are compiled using options
"\-no\-alias\-deps \-w \-49", and that those files or their interface are
passed with the "\-map" option when computing dependencies for other
files. Note also that for dependencies to be correct in the
implementation of a map file, its interface should not coerce any of
the aliases it contains.
.TP
.B \-debug\-map
Dump the delayed dependency map for each map file.
.TP
.BI \-I " directory"
Add the given directory to the list of directories searched for
source files. If a source file foo.ml mentions an external
compilation unit Bar, a dependency on that unit's interface
bar.cmi is generated only if the source for bar is found in the
current directory or in one of the directories specified with
.BR \-I .
Otherwise, Bar is assumed to be a module from the standard library,
and no dependencies are generated. For programs that span multiple
directories, it is recommended to pass
.BR ocamldep (1)
the same
.B \-I
options that are passed to the compiler.
.TP
.B \-nocwd
Do not add current working directory to the list of include directories.
.TP
.BI \-impl " file"
Process
.IR file
as a .ml file.
.TP
.BI \-intf " file"
Process
.IR file
as a .mli file.
.TP
.BI \-keywords " version+list"
Set keywords according to the
.IR version+list
specification.
.TP
.BI \-map " file"
Read an propagate the delayed dependencies for module aliases in
.IR file ,
so that the following files will depend on the
exported aliased modules if they use them.
.TP
.BI \-ml\-synonym " .ext"
Consider the given extension (with leading dot) to be a synonym for .ml.
.TP
.BI \-mli\-synonym " .ext"
Consider the given extension (with leading dot) to be a synonym for .mli.
.TP
.B \-modules
Output raw dependencies of the form
.IR filename : " Module1 Module2" " ... " ModuleN
where
.IR Module1 ", ...," " ModuleN"
are the names of the compilation
units referenced within the file
.IR filename ,
but these names are not
resolved to source file names.  Such raw dependencies cannot be used
by
.BR make (1),
but can be post-processed by other tools such as
.BR Omake (1).
.TP
.BI \-native
Generate dependencies for a pure native-code program (no bytecode
version).  When an implementation file (.ml file) has no explicit
interface file (.mli file),
.BR ocamldep (1)
generates dependencies on the
bytecode compiled file (.cmo file) to reflect interface changes.
This can cause unnecessary bytecode recompilations for programs that
are compiled to native-code only.  The flag
.B \-native
causes dependencies on native compiled files (.cmx) to be generated instead
of on .cmo files.  (This flag makes no difference if all source files
have explicit .mli interface files.)
.TP
.B \-one-line
Output one line per file, regardless of the length.
.TP
.BI \-open " module"
Assume that module
.IR module
is opened before parsing each of the
following files.
.TP
.BI \-pp " command"
Cause
.BR ocamldep (1)
to call the given
.I command
as a preprocessor for each source file.
.TP
.BI \-ppx " command"
Pipe abstract syntax tree through preprocessor
.IR command .
.TP
.B \-shared
Generate dependencies for native plugin files (.cmxs) in addition to
native compiled files (.cmx).
.TP
.B \-slash
(Windows) Use forward slash / instead of backslash \\ in file paths.
Under Unix, this option does nothing.
.TP
.B \-no\-slash
(Windows) Preserve any backslash \\ in file paths.
Under Unix, this option does nothing.
.TP
.B \-sort
Sort files according to their dependencies.
.TP
.B \-version
Print version string and exit.
.TP
.B \-vnum
Print short version number and exit.
.TP
.BI \-args " file"
Read additional newline separated command line arguments from
.IR file .
.TP
.BI \-args0 " file"
Read additional NUL separated command line arguments from
.IR file .
.TP
.BR \-help " or " \-\-help
Display a short usage summary and exit.

.SH SEE ALSO
.BR ocamlc "(1), " ocamlopt (1).
.br
.IR The\ OCaml\ user's\ manual ,
chapter "Dependency generator".


News
OCaml 4.10.0 (21 February 2020)
-------------------------------

- New best-fit allocator for the major heap
- Preliminary runtime work for OCaml multicore
- Immutable strings are now enforced at configuration time
- User-defined indexing operators for multidimensional arrays
- Coming soon: statmemprof, a new statistical memory profiler.
- The external API will be released next version.
- Various improvements to the manual
- More precise exhaustiveness check for GADTs
- Many bug fixes


OCaml 4.09.1 (18 March 2020)
----------------------------

Bug fixes.

OCaml 4.09.0 (18 September 2019)
--------------------------------

- New optimisations, in particular for affine functions in matches, for instance:

     type t = A | B | C
     let affine = function
       | A -> 4
       | B -> 3
       | C -> 2

- The `graphics` library was moved out of the compiler distribution.
- The `vmthread` library was removed.
- Support for compiler plugins was removed.
- Many bug fixes.

OCaml 4.08.1 (5 August 2019)
----------------------------

Bug fixes.

OCaml 4.08.0 (14 June 2019)
---------------------------

- Binding operators (let*, let+, and*, etc). They can be used to
  streamline monadic code.

- `open` now applies to arbitrary module expression in structures and
  to applicative paths in signatures.

- A new notion of (user-defined) "alerts" generalizes the deprecated
  warning.

- New modules in the standard library: Fun, Bool, Int, Option, Result.

- A significant number of new functions in Float, including FMA
  support, and a new Float.Array submodule.

- Source highlighting for errors and warnings in batch mode.

- Many error messages were improved.

- Improved AFL instrumentation for objects and lazy values.


OCaml 4.07.1 (4 October 2018)
-----------------------------

This release consists mostly of bug fixes. The most salient bugs were

- MPR#7820, GPR#1897: a bug in Array.of_seq (new in 4.07)
  (Thierry Martinez, review by Nicolás Ojeda Bär)

- MPR#7815, GPR#1896: crash in the major GC with the first-fit policy
  (Stephen Dolan and Damien Doligez, report by Joris Giovannangeli)

- MPR#7821, GPR#1908: the compiler loads more cmi, which breaks some builds
  (Jérémie Dimino, review by Gabriel Scherer)

- MPR#7833, GPR#1946: typechecking failure (regression) on large GADT matchings
  (Thomas Refis, report by Jerome Simeon, review by Jacques Garrigue)

See the detailed list of fixes at (Changes#4.07.1).


OCaml 4.07.0 (10 July 2018):
----------------------------

Some highlights of this release are:

-   The way the standard library modules are organized internally has
    changed (GPR#1010, by Jérémie Dimino):

    1. the `List` module (for example) is now named `Stdlib__list`
    2. a new Stdlib module contains a series of aliases
    such as `module List = Stdlib__list`
    3. the `Stdlib` module is implicitly opened when type-checking OCaml
    programs (as `Pervasives` previously was), so that `Stdlib.List` can be
    accessed as just `List`, as before.

    This should be invisible to most users, although it is possible that
    some tools show the `Stdlib.` or `Stdlib__` prefixes in
    messages. (You might want to report these situations as usability
    bugs.) The change prevents standard library modules from conflicting
    with end-user filenames (please avoid `stdlib.ml` and the
    `Stdlib__` prefix); we may introduce new standard library modules in
    the future with less fear of breaking user code. In particular,
    `Float` (GPR#1638, by Nicolás Ojeda Bär) and `Seq` (GPR#1002, by
    Simon Cruanes) modules have now been added to the standard library.

-   The error messages caused by various typing errors have been improved
    to be easier to understand, in particular for beginners.
    (GPR#1505, GPR#1510, by Arthur Charguéreau and Armaël Guéneau)

    For example,

        # while 1 do () done;;
                ^
        Error: This expression has type int but
               an expression was expected of type bool

    now adds the extra explanation

        because it is in the condition of a while-loop

-   Effort has been made to reduce the compilation time of flambda
    programs, and the size of the produced `.cmx` files when using
    the -Oclassic optimisation level.
    (GPR#1401, GPR#1455, GPR#1627, GPR#1665, by Pierre Chambart, Xavier
    Clerc, Fuyong Quah, and Leo White)

-   The HTML manual has benefited from various style improvements
    and should look visually nicer than previous editions.
    (GPR#1741, GPR#1757, GPR#1767 by Charles Chamberlain and steinuil)

    The new version of the manual can be consulted at
    <http://caml.inria.fr/pub/docs/manual-ocaml-4.07/>; see the
    previous version for comparison at
    <http://caml.inria.fr/pub/docs/manual-ocaml-4.06/>.

-   Since 4.01, it is possible to select a variant constructor or
    record field from a sub-module that is not opened in the current
    scope, if type information is available at the point of use. This
    now also works for GADT constructors.
    (GPR#1648, by Thomas Refis and Leo White)

-   The GC should handle the accumulation of custom blocks in the minor
    heap better; this solves some memory-usage issues observed by code
    which allocates a lot of small custom blocks, typically small bigarrays
    (GPR#1476, by Alain Frsich)

See also the detailed list of changes: (Changes#4.07.0).


OCaml 4.06.1 (16 Feb 2018):
---------------------------

This release consists mostly of bug fixes. The most salient bugs were

-   An incorrect compilation of pattern-matching in presence of
    extensible variant constructors (such as exceptions), that had been
    present for a long time.
    (GPR#1459, GPR#1538, by Luc Maranget, Thomas Refis and Gabriel Scherer)

-   An optimization of `not (x = y)` into `x <> y`, introduced in
    4.06.0, is incorrect on floating-point numbers in the `nan`
    case. (GPR#1470, by Leo White)

See the detailed list of fixes at (Changes#4.06.1).



man/ocamldep.m
.\"**************************************************************************
.\"*                                                                        *
.\"*                                 OCaml                                  *
.\"*                                                                        *
.\"*             Xavier Leroy, projet Cristal, INRIA Rocquencourt           *
.\"*                                                                        *
.\"*   Copyright 1996 Institut National de Recherche en Informatique et     *
.\"*     en Automatique.                                                    *
.\"*                                                                        *
.\"*   All rights reserved.  This file is distributed under the terms of    *
.\"*   the GNU Lesser General Public License version 2.1, with the          *
.\"*   special exception on linking described in the file LICENSE.          *
.\"*                                                                        *
.\"**************************************************************************
.\"
.TH OCAMLDEP 1

.SH NAME
ocamldep \- Dependency generator for OCaml

.SH SYNOPSIS
.B ocamldep
[
.I options
]
.I filename ...

.SH DESCRIPTION

The
.BR ocamldep (1)
command scans a set of OCaml source files
(.ml and .mli files) for references to external compilation units,
and outputs dependency lines in a format suitable for the
.BR make (1)
utility. This ensures that make will compile the source files in the
correct order, and recompile those files that need to when a source
file is modified.

The typical usage is:
.P
ocamldep
.I options
*.mli *.ml > .depend
.P
where .depend is the file that should contain the
dependencies.

Dependencies are generated both for compiling with the bytecode
compiler
.BR ocamlc (1)
and with the native-code compiler
.BR ocamlopt (1).

.SH OPTIONS

The following command-line options are recognized by
.BR ocamldep (1).
.TP
.B \-absname
Show absolute filenames in error messages.
.TP
.B \-all
Generate dependencies on all required files, rather than assuming
implicit dependencies.
.TP
.B \-allow\-approx
Allow falling back on a lexer-based approximation when parsing fails.
.TP
.B \-as\-map
For the following files, do not include delayed dependencies for
module aliases.
This option assumes that they are compiled using options
"\-no\-alias\-deps \-w \-49", and that those files or their interface are
passed with the "\-map" option when computing dependencies for other
files. Note also that for dependencies to be correct in the
implementation of a map file, its interface should not coerce any of
the aliases it contains.
.TP
.B \-debug\-map
Dump the delayed dependency map for each map file.
.TP
.BI \-I \ directory
Add the given directory to the list of directories searched for
source files. If a source file foo.ml mentions an external
compilation unit Bar, a dependency on that unit's interface
bar.cmi is generated only if the source for bar is found in the
current directory or in one of the directories specified with
.BR \-I .
Otherwise, Bar is assumed to be a module from the standard library,
and no dependencies are generated. For programs that span multiple
directories, it is recommended to pass
.BR ocamldep (1)
the same
.B \-I
options that are passed to the compiler.
.TP
.B \-nocwd
Do not add current working directory to the list of include directories.
.TP
.BI \-impl \ file
Process
.IR file
as a .ml file.
.TP
.BI \-intf \ file
Process
.IR file
as a .mli file.
.TP
.BI \-map \ file
Read an propagate the delayed dependencies for module aliases in
.IR file ,
so that the following files will depend on the
exported aliased modules if they use them.
.TP
.BI \-ml\-synonym \ .ext
Consider the given extension (with leading dot) to be a synonym for .ml.
.TP
.BI \-mli\-synonym \ .ext
Consider the given extension (with leading dot) to be a synonym for .mli.
.TP
.B \-modules
Output raw dependencies of the form
.IR filename : \ Module1\ Module2 \ ... \ ModuleN
where
.IR Module1 ,\ ..., \ ModuleN
are the names of the compilation
units referenced within the file
.IR filename ,
but these names are not
resolved to source file names.  Such raw dependencies cannot be used
by
.BR make (1),
but can be post-processed by other tools such as
.BR Omake (1).
.TP
.B \-json
Print module dependencies in JSON format only supported with -modules for now.
.TP
.BI \-native
Generate dependencies for a pure native-code program (no bytecode
version).  When an implementation file (.ml file) has no explicit
interface file (.mli file),
.BR ocamldep (1)
generates dependencies on the
bytecode compiled file (.cmo file) to reflect interface changes.
This can cause unnecessary bytecode recompilations for programs that
are compiled to native-code only.  The flag
.B \-native
causes dependencies on native compiled files (.cmx) to be generated instead
of on .cmo files.  (This flag makes no difference if all source files
have explicit .mli interface files.)
.TP
.B \-one-line
Output one line per file, regardless of the length.
.TP
.BI \-open \ module
Assume that module
.IR module
is opened before parsing each of the
following files.
.TP
.BI \-pp \ command
Cause
.BR ocamldep (1)
to call the given
.I command
as a preprocessor for each source file.
.TP
.BI \-ppx \ command
Pipe abstract syntax tree through preprocessor
.IR command .
.TP
.B \-shared
Generate dependencies for native plugin files (.cmxs) in addition to
native compiled files (.cmx).
.TP
.B \-slash
Under Unix, this option does nothing.
.TP
.B \-sort
Sort files according to their dependencies.
.TP
.B \-version
Print version string and exit.
.TP
.B \-vnum
Print short version number and exit.
.TP
.BR \-help \ or \ \-\-help
Display a short usage summary and exit.

.SH SEE ALSO
.BR ocamlc (1),
.BR ocamlopt (1).
.br
.IR The\ OCaml\ user's\ manual ,
chapter "Dependency generator".
